src/app/components/client/toolbar/AnnouncementDialog.tsx

```diff
@@ -0,0 +1,404 @@
+import { LocalizedAnnouncementString } from "../../../(proper_react)/(redesign)/(authenticated)/admin/announcements/AnnouncementsAdmin";
+import { TelemetryLink } from "../TelemetryLink";
+import { useTelemetry } from "../../../hooks/useTelemetry";
+
+type AnnouncementDialogProps = {
+  announcements: UserAnnouncementWithDetails[];
+};
+
+export const AnnouncementDialog = ({
+  announcements,
+}: AnnouncementDialogProps) => {
+  const l10n = useL10n();
+  const recordTelemetry = useTelemetry();
+  const triggerRef = useRef<HTMLButtonElement>(null);
+  const triggerState = useOverlayTriggerState({
+    onOpenChange: (isOpen) => {
+      if (isOpen) {
+        recordTelemetry("button", "click", {
+          button_id: "opened_announcements",
+        });
+      }
+    },
+  });
+  const { triggerProps, overlayProps } = useOverlayTrigger(
+    { type: "dialog" },
+    triggerState,
+    triggerRef,
+  );
+  const { buttonProps } = useButton(triggerProps, triggerRef);
+  const dismissButtonRef = useRef<HTMLButtonElement>(null);
+  const dismissButtonProps = useButton(
+    {
+      /* c8 ignore next 3 */
+      onPress: () => {
+        triggerState.close();
+      },
+    },
+    dismissButtonRef,
+  ).buttonProps;
+
+  const popoverRef = useRef(null);
+  const [smallImageUnavailableMap, setSmallImageUnavailableMap] = useState<
+    Record<string, boolean>
+  >({});
+  const [bigImageUnavailableMap, setBigImageUnavailableMap] = useState<
+    Record<string, boolean>
+  >({});
+  const [announcementDetailsView, setAnnouncementDetailsView] = useState(false);
+  const [activeTab, setActiveTab] = useState<"new" | "all">("new");
+
+  const [relevantAnnouncement, setRelevantAnnouncement] =
+    useState<UserAnnouncementWithDetails | null>(null);
+
+  const [userAnnouncements, setUserAnnouncements] =
+    useState<UserAnnouncementWithDetails[]>(announcements);
+
+  const sortedAnnouncements = [...userAnnouncements].sort((a, b) => {
+    return new Date(b.created_at).getTime() - new Date(a.created_at).getTime();
+  });
+
+  const numNewAnnouncements = userAnnouncements.filter(
+    (a) => a.status === "new",
+  ).length;
+
+  const filteredAnnouncements = sortedAnnouncements.filter((a) =>
+    activeTab === "new"
+      ? a.status === "new"
+      : a.status === "seen" || a.status === "cleared",
+  );
+
+  const handleMarkAsSeen = async (
+    announcement: UserAnnouncementWithDetails,
+  ) => {
+    // Optimistically handle state change
+    setUserAnnouncements((prev) =>
+      prev.map((a) =>
+        a.announcement_id === announcement.announcement_id
+          ? {
+              ...a,
+              status: "seen",
+              clicked_at: new Date(),
+              seen_at: new Date(),
+            }
+          : a,
+      ),
+    );
+
+    try {
+      await fetch(
+        `/api/v1/user/announcements/${announcement.announcement_id}/seen`,
+        {
+          method: "PUT",
+        },
+      );
```

Comment:
One general question: @rhelmer @codemist I remember we’ve discussed before that we are trying to use Server Actions when possible. For the announcements API calls, we are making requests through `/api/v1` here and an in other places.